tests/std/tests/GH_002030_asan_annotate_vector/test.cpp

```diff
@@ -274,6 +274,93 @@ struct implicit_allocator : custom_test_allocator<T, Pocma, Stateless> {
 STATIC_ASSERT(_Container_allocation_minimum_asan_alignment<vector<char, implicit_allocator<char>>> == 1);
 STATIC_ASSERT(_Container_allocation_minimum_asan_alignment<vector<wchar_t, implicit_allocator<wchar_t>>> == 2);
 
+// Helper class for `ArenaAllocator`, where data is linearly allocated in a finite buffer.
+class Arena {
+public:
+    Arena(size_t allocation_size, size_t capacity):
+        data(new char[capacity* allocation_size]),
+        allocation_size(allocation_size),
+        capacity(capacity),
+        offset(0) {}
+
+    Arena(const Arena& other):
+        data(other.data),
+        allocation_size(other.allocation_size),
+        capacity(other.capacity),
+        offset(other.offset) {}
+
+    void* allocate(size_t n) {
+        assert(offset + n <= capacity && "Allocation failed: the arena is out of memory. You may call `reset` to free up space.");
+        void* result = data + (offset * allocation_size);
+        offset += n;
+        return result;
+    }
+
+    // Deallocates memory in the arena, and `memset`s it all to zero
+    void reset() noexcept {
+        offset = 0;
+        memset(data, 0, capacity*allocation_size);
+    }
+
+    char* const data;
+    const size_t allocation_size;
+    const size_t capacity;
+    size_t offset;
+};
+
+// An allocator that allocates memory in a pre-allocated buffer (the arena).
+template <typename T>
+class ArenaAllocator {
+public:
+
+    using value_type = T;
+    Arena* arena;
+
+    ArenaAllocator(size_t alloc_size, size_t capacity) {
+        assert(sizeof(T) <= alloc_size && "Constructor failed: allocation size is too small for target type.");
+        arena = new Arena(alloc_size, capacity);
+    }
+
+    template <class U>
+    ArenaAllocator(const ArenaAllocator<U>& other) {
+        assert(sizeof(U) <= arena->allocation_size && "Constructor failed: allocation size is too small for target type.");
+        arena = other.arena;
+    }
+
+    T* allocate(size_t n) {
+        void* ptr = arena->allocate(n);
+		return reinterpret_cast<T*>(ptr);
+    }
+
+    // no-op. Memory is deallocated in the `reset` method
+    void deallocate(value_type*, size_t) noexcept {}
+
+    // Deallocates memory in the arena, and `memset`s it all to zero.
+    // the `memset` would normally trigger an ASan AV,
+    // but we'll have the ArenaAllocator opt out of ASan analysis
+    void reset() noexcept {
+        arena->reset();
+    }
+
+    template <typename U>
+    bool operator==(ArenaAllocator<U> const& other)
+    {
+        return this->arena == other.arena;
+    }
+
+    template<typename U>
+    bool operator!=(ArenaAllocator<U> const other)
+    {
+        return !(this == other);
+    }
+};
+
+// Opt out of ASan analysis for the ArenaAllocator
+// FIXME: IntelliSense claims '_Is_ASan_enabled_for_allocator is not a templateC/C++(864)',
+//  but it works somehow.

```

Comment:
This is probably due to IntelliSense picking up the installed STL headers instead of the headers in your STL repo.

Reply:
Oh, that makes sense. I'll remove this comment in my next commit then.

Just curious - do we have a good way of teling intellisense about this, or do we just 'ignore' the false alarm?

Reply:
We test IntelliSense via the `/BE` command-line option. That should pick up your freshly built STL headers.

As for actual IntelliSense in the IDE, no idea. I would expect that if you've built something with the freshly built STL picked up (remember `set_environment.bat`!), then it would know the correct paths.

Reply:
I'll do some digging and find out, would be good to document :-)

Reply:
Resolving; comment has been removed.